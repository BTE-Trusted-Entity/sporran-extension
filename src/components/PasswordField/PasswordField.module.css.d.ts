export const __esModule: true;
export const resetLine: string;
export const reset: string;
export const passwordLabel: string;
export const passwordLine: string;
export const password: string;
export const errorTooltip: string;
export const rememberLabel: string;
export const remember: string;
