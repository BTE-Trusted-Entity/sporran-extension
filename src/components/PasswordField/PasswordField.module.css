.resetLine {
  align-self: stretch;
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 0;
}

.reset {
  composes: smallText from '../Typography/Typography.module.css';
  color: var(--color-lightest);
}

.passwordLabel {
  composes: smallText from '../Typography/Typography.module.css';
  margin: 0.75rem 0;
  text-align: left;
}

.passwordLine {
  align-self: stretch;
  position: relative;
  margin: 0;
}

.password {
  composes: input from '../Input/Input.module.css';
  width: calc(100% - 3.5rem);
  padding-right: 2.5rem;
}

.errorTooltip {
  composes: tooltip from '../ErrorTooltip/ErrorTooltip.module.css';
}

.rememberLabel {
  composes: smallText from '../Typography/Typography.module.css';
  align-self: stretch;
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 0.75rem 0.5rem 0.75rem 0;
  cursor: pointer;
}

.remember {
  composes: toggle from '../Toggle/Toggle.module.css';
}
