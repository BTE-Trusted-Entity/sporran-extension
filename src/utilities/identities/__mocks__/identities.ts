export const decryptIdentity = jest.fn();

export const saveIdentity = jest.fn();
