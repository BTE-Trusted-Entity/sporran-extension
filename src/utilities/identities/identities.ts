import { useContext } from 'react';
import { mutate } from 'swr';
import {
  blake2AsU8a,
  keyExtractPath,
  keyFromPath,
  mnemonicToMiniSecret,
  sr25519PairFromSeed,
} from '@polkadot/util-crypto';
import {
  Did,
  DidDocument,
  DidResourceUri,
  DidUri,
  EncryptRequestData,
  IEncryptedMessage,
  KiltAddress,
  KiltEncryptionKeypair,
  KiltKeyringPair,
  Message,
  MessageBody,
  SignCallback,
  SignRequestData,
  Utils,
} from '@kiltprotocol/sdk-js';
import { map, max, memoize } from 'lodash-es';

import {
  loadEncrypted,
  saveEncrypted,
} from '../storageEncryption/storageEncryption';

import { getDidDocument, getDidEncryptionKey, parseDidUri } from '../did/did';
import { storage } from '../storage/storage';
import { useSwrDataOrThrow } from '../useSwrDataOrThrow/useSwrDataOrThrow';

import { IdentitiesContext, IdentitiesContextType } from './IdentitiesContext';
import { getIdentities, IDENTITIES_KEY } from './getIdentities';

import { Identity } from './types';

export { Identity, IdentitiesMap } from './types';

const CURRENT_IDENTITY_KEY = 'currentIdentity';

export const NEW: Identity = {
  address: 'NEW' as KiltAddress,
  did: '' as DidUri,
  name: '',
  index: -1,
};

export function isNew(identity: Identity): boolean {
  return identity === NEW;
}

export function useIdentities(): IdentitiesContextType {
  return useContext(IdentitiesContext);
}

async function getCurrentIdentity(): Promise<string | null> {
  const identities = await getIdentities();
  const stored = await storage.get(CURRENT_IDENTITY_KEY);
  const current = stored[CURRENT_IDENTITY_KEY];

  if (identities[current]) {
    return current;
  }

  const firstIdentity = Object.values(identities)[0];
  if (!firstIdentity) {
    return null;
  }

  await setCurrentIdentity(firstIdentity.address);

  return firstIdentity.address;
}

export async function setCurrentIdentity(address: string): Promise<void> {
  const oldAddress = await storage.get(CURRENT_IDENTITY_KEY);
  if (address === oldAddress[CURRENT_IDENTITY_KEY]) {
    return;
  }
  await storage.set({ [CURRENT_IDENTITY_KEY]: address });
  await mutate(['getCurrentIdentity', CURRENT_IDENTITY_KEY]);
}

export async function saveIdentity(identity: Identity): Promise<void> {
  const identities = await getIdentities();
  identities[identity.address] = identity;
  await storage.set({ [IDENTITIES_KEY]: identities });
  await mutate(IDENTITIES_KEY);
}

export async function removeIdentity(identity: Identity): Promise<void> {
  const identities = await getIdentities();
  delete identities[identity.address];

  await storage.set({ [IDENTITIES_KEY]: identities });
  await storage.remove(identity.address);

  const firstIdentity = Object.values(identities)[0];
  await setCurrentIdentity(firstIdentity?.address);

  await mutate(IDENTITIES_KEY);
}

interface IdentityDidCrypto {
  didDocument: DidDocument;
  keypair: KiltKeyringPair;
  authenticationKey: KiltKeyringPair;
  sign: SignCallback;
  encrypt: (
    messageBody: MessageBody,
    dAppDidDocument: DidDocument,
  ) => Promise<IEncryptedMessage>;
}

export function deriveAuthenticationKey(seed: Uint8Array) {
  const baseKey = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  return baseKey.derive('//did//0') as typeof baseKey;
}

export function deriveEncryptionKeyFromSeed(
  seed: Uint8Array,
): KiltEncryptionKeypair {
  const keypair = sr25519PairFromSeed(seed);
  const { path } = keyExtractPath('//did//keyAgreement//0');
  const { secretKey } = keyFromPath(keypair, path, 'sr25519');
  return Utils.Crypto.makeEncryptionKeypairFromSeed(blake2AsU8a(secretKey));
}

export function deriveAttestationKeyFromSeed(seed: Uint8Array) {
  const baseKey = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  return baseKey.derive('//did//assertion//0') as typeof baseKey;
}

function deriveEncryptionKeyLegacy(seed: Uint8Array) {
  const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const encryptionKeyringPair = keypair.derive('//did//keyAgreement//0');

  const secret = encryptionKeyringPair
    .encryptMessage(
      new Uint8Array(24).fill(0),
      new Uint8Array(24).fill(0),
      new Uint8Array(24).fill(0),
    )
    .slice(24); // first 24 bytes are the nonce

  return Utils.Crypto.makeEncryptionKeypairFromSeed(secret);
}

async function fixLightDidIssues(seed: Uint8Array) {
  const identities = await getIdentities();
  const { address } = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const identity = identities[address];

  if (!identity) {
    // could be the Alice identity
    return;
  }

  const parsed = identity.did && Did.parse(identity.did);
  if (parsed && parsed.type !== 'light') {
    return;
  }

  try {
    if (!identity.did) {
      // DID was deactivated, no action needed.
      return;
    }

    // If this light DID was created and stored using SDK@0.24.0 then its keys are serialized using base64,
    // resulting in an invalid URI, so resolving would throw an exception.
    const document = await getDidDocument(identity.did);

    // Another issue we see is the light DIDs without key agreement keys, need to regenerate them as well
    const encryptionKey = getDidEncryptionKey(document);

    // This public key also means the DID needs to be regenerated
    const troubleKey =
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e';
    if (Utils.Crypto.u8aToHex(encryptionKey.publicKey) === troubleKey) {
      throw new Error();
    }
  } catch {
    // We re-create the invalid DID from scratch and update its URI in the identity.
    const { uri } = getLightDidFromSeed(seed);
    await saveIdentity({ ...identity, did: uri });
  }
}

export async function getIdentityCryptoFromSeed(
  seed: Uint8Array,
  legacy?: boolean,
): Promise<IdentityDidCrypto> {
  await fixLightDidIssues(seed);

  const authenticationKey = deriveAuthenticationKey(seed);
  const encryptionKey = legacy
    ? deriveEncryptionKeyLegacy(seed)
    : deriveEncryptionKeyFromSeed(seed);

  const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const identities = await getIdentities();
  const did = getIdentityDid(identities[keypair.address]);

  const didDocument = await getDidDocument(did);

  async function encryptCallback({ data, peerPublicKey }: EncryptRequestData) {
    const { box, nonce } = Utils.Crypto.encryptAsymmetric(
      data,
      peerPublicKey,
      encryptionKey.secretKey,
    );
    const keyUri = `${didDocument.uri}${getDidEncryptionKey(didDocument).id}`;
    return {
      nonce,
      data: box,
      keyUri: keyUri as DidResourceUri,
    };
  }

  async function sign({ data, keyRelationship }: SignRequestData) {
    if (keyRelationship !== 'authentication') {
      throw new Error(
        'Only key relationship "authentication" is supported for signing',
      );
    }

    const signature = authenticationKey.sign(data, { withType: false });
    const keyUri =
      `${didDocument.uri}${didDocument.authentication[0].id}` as DidResourceUri;
    const keyType = authenticationKey.type;

    return {
      signature,
      keyUri,
      keyType,
    };
  }

  async function encrypt(
    messageBody: MessageBody,
    dAppDidDocument: DidDocument,
  ): Promise<IEncryptedMessage> {
    const message = Message.fromBody(
      messageBody,
      didDocument.uri,
      dAppDidDocument.uri,
    );

    return Message.encrypt(
      message,
      encryptCallback,
      `${dAppDidDocument.uri}${getDidEncryptionKey(dAppDidDocument).id}`,
    );
  }

  return {
    didDocument,
    keypair,
    authenticationKey,
    sign,
    encrypt,
  };
}

export async function encryptIdentity(
  backupPhrase: string,
  password: string,
): Promise<KiltAddress> {
  const seed = mnemonicToMiniSecret(backupPhrase);
  const { address } = Utils.Crypto.makeKeypairFromUri(backupPhrase, 'sr25519');
  await saveEncrypted(address, password, seed);
  return address;
}

export function getLightDidFromSeed(seed: Uint8Array): DidDocument {
  return Did.createLightDidDocument({
    authentication: [deriveAuthenticationKey(seed)],
    keyAgreement: [deriveEncryptionKeyFromSeed(seed)],
  });
}

async function getIdentityName(): Promise<{ name: string; index: number }> {
  const identities = await getIdentities();
  const largestIndex = max(map(identities, 'index')) || 0;

  const index = 1 + largestIndex;

  return { name: `KILT Identity ${index}`, index };
}

export async function createIdentity(
  backupPhrase: string,
  password: string,
): Promise<Identity> {
  const address = await encryptIdentity(backupPhrase, password);

  const seed = mnemonicToMiniSecret(backupPhrase);

  const { uri } = getLightDidFromSeed(seed);

  const { name, index } = await getIdentityName();

  const identity = { name, address, did: uri, index };
  await saveIdentity(identity);

  return identity;
}

export async function importIdentity(
  backupPhrase: string,
  password: string,
): Promise<Identity> {
  const address = await encryptIdentity(backupPhrase, password);

  const seed = mnemonicToMiniSecret(backupPhrase);

  const lightDidDocument = getLightDidFromSeed(seed);
  const lightDid = lightDidDocument.uri;

  const resolved = await Did.resolve(lightDid);
  const fullDid = resolved?.metadata?.canonicalId;

  const deactivated = resolved?.metadata?.deactivated;
  const did = deactivated ? undefined : fullDid || lightDid;
  const deletedDid = deactivated ? Did.getFullDidUri(lightDid) : undefined;

  const { name, index } = await getIdentityName();

  const identity = {
    name,
    address,
    index,
    ...(!deactivated && { did }),
    ...(deactivated && { deletedDid }),
  };
  await saveIdentity(identity);

  return identity;
}

export async function decryptIdentity(
  address: string,
  password: string,
): Promise<Uint8Array> {
  const seed = await loadEncrypted(address, password);
  return new Uint8Array(seed);
}

/** Ensure that local information about the DID type matches stored on blockchain
 * even if an error occurred while asynchronous update was running */
async function syncDidStateWithBlockchain(address: string | null | undefined) {
  if (!address) {
    return;
  }

  const identities = await getIdentities();
  const identity = identities[address];

  if (!identity.did) {
    // could be a legacy identity without DID
    return;
  }

  const { fullDid, type } = parseDidUri(identity.did);
  const wasOnChain = type === 'full';

  const resolved = await Did.resolve(identity.did);
  const isOnChain = wasOnChain
    ? Boolean(resolved && resolved.metadata && !resolved.metadata.deactivated)
    : Boolean(resolved && resolved.metadata && resolved.metadata.canonicalId);

  if (wasOnChain && !isOnChain) {
    await saveIdentity({
      ...identity,
      did: undefined,
      deletedDid: fullDid,
    });
  }
  if (!wasOnChain && isOnChain) {
    await saveIdentity({ ...identity, did: fullDid });
  }
}

/** Memoized function will run only once per identity while the popup is open, do not use from backend */
const noAwaitUpdateCachedDidStateOnce = memoize(syncDidStateWithBlockchain);

export function useCurrentIdentity(): string | null | undefined {
  const data = useSwrDataOrThrow(
    CURRENT_IDENTITY_KEY,
    getCurrentIdentity,
    'getCurrentIdentity',
  );

  noAwaitUpdateCachedDidStateOnce(data);

  return data;
}

export function getIdentityDid({ did }: Identity): DidUri {
  if (!did) {
    throw new Error('DID is deleted and unusable');
  }
  return did;
}
