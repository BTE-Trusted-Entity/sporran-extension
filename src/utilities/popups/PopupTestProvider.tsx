import { MemoryRouter } from 'react-router-dom';

import { jsonToBase64 } from '../base64/base64';

interface Props {
  path: string;
  data: unknown;
  children: JSX.Element;
}

export function PopupTestProvider({
  path,
  data,
  children,
}: Props): JSX.Element {
  const serialized = jsonToBase64(data);
  return (
    <MemoryRouter initialEntries={[`${path}?data=${serialized}`]}>
      {children}
    </MemoryRouter>
  );
}
