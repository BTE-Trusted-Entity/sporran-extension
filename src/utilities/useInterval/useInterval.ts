import { useEffect } from 'react';

type IntervalFunction = () => unknown | void;

export function useInterval(callback: IntervalFunction, delay: number): void {
  useEffect(() => {
    const id = setInterval(callback, delay);
    return () => clearInterval(id);
  }, [callback, delay]);
}
