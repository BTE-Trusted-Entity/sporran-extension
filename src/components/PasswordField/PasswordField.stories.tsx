import { Meta } from '@storybook/react';

import { identitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { PasswordField, usePasswordField } from './PasswordField';

export default {
  title: 'Components/PasswordField',
  component: PasswordField,
} as Meta;

const identity = identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'];

export function Template(): JSX.Element {
  const passwordField = usePasswordField();
  return (
    <form>
      <PasswordField identity={identity} password={passwordField} />
    </form>
  );
}
