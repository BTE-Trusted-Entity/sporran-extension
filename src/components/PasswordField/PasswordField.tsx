import {
  Dispatch,
  FormEvent,
  Fragment,
  SetStateAction,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';
import { KiltKeyringPair, Utils } from '@kiltprotocol/sdk-js';

import * as styles from './PasswordField.module.css';

import { decryptIdentity } from '../../utilities/identities/identities';
import { usePasswordType } from '../usePasswordType/usePasswordType';
import { useInterval } from '../../utilities/useInterval/useInterval';
import {
  forgetPasswordChannel,
  getPasswordChannel,
  savePasswordChannel,
} from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels';
import { RouteExcept } from '../RouteExcept/RouteExcept';
import { generatePath, paths } from '../../views/paths';
import { PasswordError } from '../../utilities/storageEncryption/storageEncryption';

export { PasswordError } from '../../utilities/storageEncryption/storageEncryption';

// Okay, ESLint, I must have a parameter but cannot use it
// eslint-disable-next-line @typescript-eslint/no-unused-vars
async function defaultGetPassword(event: FormEvent): Promise<Value> {
  throw new Error('Not initialized yet, getPassword not set');
}

// To store a function in state I have to use this workaround since proper typing was not provided
// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38160
type SetPasswordGetterType = Dispatch<
  SetStateAction<(event: FormEvent) => Promise<Value>>
>;

interface Value {
  password: string;
  keypair: KiltKeyringPair;
  seed: Uint8Array;
}

export function usePasswordField(): {
  get: (event: FormEvent) => Promise<Value>;
  set: (getter: (event: FormEvent) => Promise<Value>) => void;
} {
  const [passwordGetter, setPasswordGetter] = useState(
    () => defaultGetPassword,
  );

  return useMemo(
    () => ({
      get: passwordGetter,
      set: setPasswordGetter,
    }),
    [passwordGetter],
  );
}

const asterisks = '************';

interface Props {
  identity: { address: string };
  autoFocus?: boolean;
  password: {
    set: (getter: (event: FormEvent) => Promise<Value>) => void;
  };
}

export function PasswordField({
  identity: { address },
  autoFocus,
  password,
}: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const setPasswordGetter = password.set as SetPasswordGetterType;

  const rememberRef = useRef<HTMLInputElement>(null);

  const [savedPassword, setSavedPassword] = useState<string | undefined>();

  const [error, setError] = useState<string | null>(null);

  const passwordGetter = useCallback(
    async (event: FormEvent) => {
      if (!address) {
        throw new Error('No identity address');
      }

      const target = event.target as unknown as {
        elements: Record<string, HTMLInputElement>;
      };

      const providedPassword = target.elements.password.value;
      const useSaved = savedPassword && providedPassword === asterisks;
      const password = useSaved ? savedPassword : providedPassword;

      if (!password) {
        setError(t('component_PasswordField_required'));
        throw new PasswordError('No password');
      }

      let seed: Uint8Array;
      try {
        seed = await decryptIdentity(address, password);
      } catch (exception) {
        if (exception instanceof PasswordError) {
          setError(t('component_PasswordField_password_incorrect'));
        }
        throw exception;
      }

      if (rememberRef.current?.checked) {
        await savePasswordChannel.get({ password, address });
      } else {
        await forgetPasswordChannel.get(address);
      }

      const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
      return { password, keypair, seed };
    },
    [address, rememberRef, savedPassword, t],
  );

  const handlePasswordInput = useCallback(() => {
    setError(null);
    setPasswordGetter(() => passwordGetter);
  }, [passwordGetter, setPasswordGetter]);

  const getSavedPassword = useCallback(async () => {
    if (!address || !rememberRef.current) {
      return;
    }
    const passwordString = await getPasswordChannel.get(address);
    setSavedPassword(passwordString);

    rememberRef.current.checked = Boolean(passwordString);

    setPasswordGetter(() => passwordGetter);
  }, [address, passwordGetter, setPasswordGetter]);

  useInterval(getSavedPassword, 1 * 60 * 1000);

  useEffect(() => {
    getSavedPassword();
  }, [getSavedPassword]);

  const { passwordType, passwordToggle } = usePasswordType();

  return (
    <Fragment>
      <p className={styles.resetLine}>
        <label className={styles.passwordLabel} htmlFor="password">
          {t('component_PasswordField_password')}
        </label>

        <RouteExcept path={paths.popup.base}>
          <Link
            to={generatePath(paths.identity.reset.start, { address })}
            className={styles.reset}
          >
            {t('component_PasswordField_reset')}
          </Link>
        </RouteExcept>
      </p>

      <p className={styles.passwordLine}>
        <input
          type={passwordType}
          onInput={handlePasswordInput}
          id="password"
          name="password"
          className={styles.password}
          defaultValue={savedPassword ? asterisks : undefined}
          autoFocus={autoFocus}
        />
        {passwordToggle}

        <output className={styles.errorTooltip} hidden={!error}>
          {error}
        </output>
      </p>

      <label className={styles.rememberLabel}>
        <span>{t('component_PasswordField_remember')}</span>
        <input
          type="checkbox"
          name="remember"
          className={styles.remember}
          ref={rememberRef}
        />
        <span />
      </label>
    </Fragment>
  );
}
