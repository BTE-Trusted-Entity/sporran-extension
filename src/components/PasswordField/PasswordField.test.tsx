import { FormEvent } from 'react';
import userEvent from '@testing-library/user-event';
import { KiltKeyringPair, Utils } from '@kiltprotocol/sdk-js';

import {
  identitiesMock as identities,
  act,
  render,
  screen,
  waitFor,
} from '../../testing/testing';
import { paths } from '../../views/paths';
import {
  forgetPasswordChannel,
  getPasswordChannel,
  savePasswordChannel,
} from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { decryptIdentity } from '../../utilities/identities/identities';
import { exceptionToError } from '../../utilities/exceptionToError/exceptionToError';
import { useInterval } from '../../utilities/useInterval/useInterval';

import { PasswordField } from './PasswordField';

jest
  .mocked(Utils.Crypto.makeKeypairFromSeed)
  .mockReturnValue({} as KiltKeyringPair);

jest.mock('../../channels/SavedPasswordsChannels/SavedPasswordsChannels');
jest.mock('../../utilities/identities/identities');

const identity = identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'];

const passwordField = {
  set: jest.fn(),
};

jest.mock('../../utilities/useInterval/useInterval');
jest.mocked(useInterval).mockImplementation(() => {
  /* do nothing */
});

describe('PasswordField', () => {
  beforeEach(() => {
    passwordField.set.mockReset();
    jest.mocked(decryptIdentity).mockReset();
    jest.mocked(getPasswordChannel.get).mockReset();
    jest.mocked(savePasswordChannel.get).mockReset();
    jest.mocked(forgetPasswordChannel.get).mockReset();
  });

  it('should match the snapshot', async () => {
    const { container } = render(
      <form>
        <PasswordField identity={identity} password={passwordField} />
      </form>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should skip the reset link in popups', async () => {
    render(
      <PopupTestProvider path={paths.popup.base} data={{}}>
        <form>
          <PasswordField identity={identity} password={passwordField} />
        </form>
      </PopupTestProvider>,
    );

    expect(screen.queryByText('Reset Password')).not.toBeInTheDocument();
  });

  it('should indicate saved password', async () => {
    const promise = Promise.resolve('password');
    jest.mocked(getPasswordChannel.get).mockReturnValue(promise);

    render(
      <form>
        <PasswordField identity={identity} password={passwordField} />
      </form>,
    );

    await act(async () => {
      await promise;
    });
    expect(await screen.findByLabelText(/Remember password/)).toBeChecked();
    expect(await screen.findByLabelText(/Sign with password/)).toHaveValue(
      '************',
    );
  });
  it('should clear the password from state if it has been cleared from background memory', async () => {
    const promise = Promise.resolve('password');
    jest.mocked(getPasswordChannel.get).mockReturnValue(promise);

    render(
      <form>
        <PasswordField identity={identity} password={passwordField} />
      </form>,
    );

    await act(async () => {
      await getPasswordChannel.get('');
    });

    expect(await screen.findByLabelText(/Remember password/)).toBeChecked();
    expect(await screen.findByLabelText(/Sign with password/)).toHaveValue(
      '************',
    );

    jest.mocked(getPasswordChannel.get).mockResolvedValue(undefined);

    const callback = jest.mocked(useInterval).mock.calls.pop()?.[0];
    callback?.();

    await act(async () => {
      await getPasswordChannel.get('');
    });

    expect(await screen.findByLabelText(/Remember password/)).not.toBeChecked();
    expect(await screen.findByLabelText(/Sign with password/)).toHaveValue('');
  });

  describe('password getter', () => {
    it('should report an invalid password', async () => {
      jest
        .mocked(decryptIdentity)
        .mockRejectedValue(new Error('Invalid password'));
      let error = '';

      async function handleSubmit(event: FormEvent) {
        event.preventDefault();
        const getPassword = passwordField.set.mock.calls.pop()[0]();
        try {
          await getPassword(event);
        } catch (exception) {
          error = exceptionToError(exception).message;
        }
      }

      render(
        <form onSubmit={handleSubmit}>
          <PasswordField identity={identity} password={passwordField} />
          <button type="submit">Submit</button>
        </form>,
      );

      await userEvent.click(await screen.findByText('Submit'));

      await waitFor(() => error !== '');
      expect(error).toEqual('No password');
    });

    it('should return the valid password', async () => {
      let password: Record<string, string> = {};

      async function handleSubmit(event: FormEvent) {
        event.preventDefault();
        const getPassword = passwordField.set.mock.calls.pop()[0]();
        password = await getPassword(event);
      }

      render(
        <form onSubmit={handleSubmit}>
          <PasswordField identity={identity} password={passwordField} />
          <button type="submit">Submit</button>
        </form>,
      );

      await userEvent.type(
        await screen.findByLabelText(/Sign with password/),
        'PASS',
      );
      await userEvent.click(await screen.findByText('Submit'));

      await waitFor(() => Object.entries(password).length);
      expect(password.password).toEqual('PASS');
    });

    it('should return the saved password and clear it', async () => {
      const promise = Promise.resolve('PASS');
      jest.mocked(getPasswordChannel.get).mockReturnValue(promise);

      let password: Record<string, string> = {};

      async function handleSubmit(event: FormEvent) {
        event.preventDefault();
        const getPassword = passwordField.set.mock.calls.pop()[0]();
        password = await getPassword(event);
      }

      render(
        <form onSubmit={handleSubmit}>
          <PasswordField identity={identity} password={passwordField} />
          <button type="submit">Submit</button>
        </form>,
      );

      await act(async () => {
        await promise;
      });
      await userEvent.click(await screen.findByLabelText(/Remember password/));
      await userEvent.click(await screen.findByText('Submit'));

      await waitFor(() => Object.entries(password).length);
      expect(password.password).toEqual('PASS');
      expect(forgetPasswordChannel.get).toHaveBeenCalledWith(identity.address);
    });

    it('should remember the valid password', async () => {
      let password = '';

      async function handleSubmit(event: FormEvent) {
        event.preventDefault();
        const getPassword = passwordField.set.mock.calls.pop()[0]();
        password = await getPassword(event);
      }

      render(
        <form onSubmit={handleSubmit}>
          <PasswordField identity={identity} password={passwordField} />
          <button type="submit">Submit</button>
        </form>,
      );

      await userEvent.type(
        await screen.findByLabelText(/Sign with password/),
        'PASS',
      );
      await userEvent.click(await screen.findByLabelText(/Remember password/));
      await userEvent.click(await screen.findByText('Submit'));

      await waitFor(() => password !== '');

      expect(savePasswordChannel.get).toHaveBeenCalledWith({
        address: identity.address,
        password: 'PASS',
      });
    });
  });
});
