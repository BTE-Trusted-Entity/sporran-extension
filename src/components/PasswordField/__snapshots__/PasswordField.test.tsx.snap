// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField should match the snapshot 1`] = `
<div>
  <form>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
  </form>
</div>
`;
